docs/src/design/decoder/main.md

```diff
@@ -0,0 +1,517 @@
+
+## Program execution
+Miden VM programs consist of a set of code blocks organized into a binary tree. The leaves of the tree contain linear sequences of instructions, and control flow is defined by the internal nodes of the tree.
+
+Managing control flow in the VM is accomplished by executing control flow operations listed in the table below. Each of these operations require exactly one VM cycle to execute.
+
+| Operation | Description |
+| --------- | ----------- |
+| `JOIN`    | Initiates processing of a new [Join block](https://hackmd.io/yr-ieh7SSKOzWw7Kdo9gnA#Join-block). |
+| `SPLIT`   | Initiates processing of a new [Split block](https://hackmd.io/yr-ieh7SSKOzWw7Kdo9gnA#Split-block). |
+| `LOOP`    | Initiates processing of a new [Loop block](https://hackmd.io/yr-ieh7SSKOzWw7Kdo9gnA#Loop-block). |
+| `REPEAT`  | Initiates a new iteration of an executing loop. |
+| `SPAN`    | Initiates processing of a new [Span block](https://hackmd.io/yr-ieh7SSKOzWw7Kdo9gnA#Span-block). |
+| `RESPAN`  | Initiates processing of a new operation batch within a span block. |
+| `END`     | Marks the end of a program block. |
+| `HALT`    | Marks the end of the entire program. |
+
+Let's consider a simple program below:
+```
+begin
+    <operations1>
+    if.true
+        <operations2>
+    else
+        <operations3>
+    end
+end
+```
+
+Block structure of this program is shown below.
+
+```
+JOIN
+    SPAN
+        <operations1>
+    END
+    SPLIT
+        SPAN
+            <operations2>
+        END
+        SPAN
+            <operations3>
+        END
+    END
+END
+```
+Executing this program on the VM can result in one of two possible instruction sequences. First, if after operations in `<operations1>` are executed the top of the stack is $1$, the VM will execute the following:
+```
+JOIN
+SPAN
+<operations1>
+END
+SPLIT
+SPAN
+<operations2>
+END
+END
+END
+HALT
+```
+However, if after `<operations1>` are executed, the top of the stack is $0$, the VM will execute the following:
+```
+JOIN
+SPAN
+<operations1>
+END
+SPLIT
+SPAN
+<operations3>
+END
+END
+END
+HALT
+```
+The main task of the decoder is to output exactly the same program hash, regardless of which one of the two possible execution paths was taken. However, before we can describe how this is achieved, we need to given an overview of the overall decoder structure.
+
+## Decoder structure
+The decoder is one of the more complex parts of the VM. It consists of the following components:
+
+* Main execution trace consisting of $24$ trace columns which contain the state of the decoder at a given cycle of a computation.
+* $3$ virtual tables (implemented via multi-set checks), which keep track of code blocks and operations executing on the VM.
+* Connection to the [hash co-processor](/miden/design/aux_table/hasher.html), which is used to offload hash computations from the decoder.
```

Comment:
Just a suggestion - it might be useful to link to the sections detailing each of these things to make it easier for people new to the design to follow and understand. For example, I believe the connection to the hash co-processor is described in the "Program block hashing" section, but that section comes before the virtual tables description. Since it's ordered differently below and not explicitly referred to in the same terms it may be harder for people unfamiliar with the architecture to follow this.

Reply:
Good point! I re-ordered the bullet points and also added links to relevant sections.